### PR TITLE
Rewrite in Rust

### PR DESCRIPTION
**Pull request by naftulikay**
Fuck Python.

#### Commands

<dl>
  <dt><code>aws-env list</code></dt>
  <dd>List all profiles in ascending order of priority with their names, profile URIs, priorities, and files.</dd>
  <dt><code>aws-env export profile_name</code></dt>
  <dd>Efficiently attempt to load the given profile by URI or name and write <code>export</code> lines to standard output.</dd>
</dl>

#### Priority

Credentials files have a non-trivial sorting order:

 - `~/.aws/credentials` should _always_ have the highest priority, so it and its profiles should always be last.
 - Credentials in `~/.aws/credentials.d/` are loaded if they either have no extension or have an extension of `.asc`, `.gpg`, `.ini`, or `.pgp`.
   - Encrypted files have a higher priority than non-encrypted files. This means that if the following all exist, `99-file.asc` will take priority above all others: `99-file`, `99-file.ini`, `99-file.asc`.
   - Other than encrypted files having higher precedence, files in `credentials.d` are sorted by their path alphabetically such that `00-something` has a lower priority than `99-something-else.ini`.

Profiles are stored in an `IndexMap<String, AwsProfile>`, so profile order within a file does matter. If two sections both have the same name in a file, the last one overwrites the first.

#### URIs

A profile URI is formed via `${file_stem:-}/${profile_name}`. If the file stem is empty and the URI begins with a `/`, look in the root credentials file.

#### Lazy Loading

When exporting credentials, lazy-loading is used to reduce the amount of work done, and hence, less operations are performed on the filesystem or via GnuPG.

Lazy-Loading strategy:

 - By URIs:
   - If file-stem is empty and the URI starts with a slash, try `~/.aws/credentials` first, then attempt URI lookup.
   - If file-stem is not empty, try loading `~/.aws/credentials.d/${file_stem}.{asc,gpg,ini,pgp}` and then attempt URI lookup.
   - If nothing is found, load all remaining files and try lookup by URI, then by profile name.
 - By Profile Name:
   - We cannot do any lazy-loading if there is no URI, we must load everything and then try to find the profile by its name in reverse sorting order so that we start with the highest priority file and work our way to the lowest.

---

### TODO

 - [x] List: List all profiles in ascending order of priority with their names, profile URIs, priorities, and files
   - [x] Table
   - [x] Table (No Header)
   - [x] Plain
   - [x] CSV
   - [x] CSV (No Header)
   - [x] JSON

 - [ ] Export
   - [ ] By profile name.
   - [ ] By URI via lazy-loading.

### CODE REVIEW

**Comment by naftulikay** [comment]
Okay so loading of all profiles is working including encrypted ones. The last bit of functionality are the actual CLI commands:

 - [ ] `aws-env list` - Load all available profiles and list them.
 - [ ] `aws-env load profile` - Load the first profile in priority-order<sup>1</sup> named `profile` and export.
 - [ ] `aws-env load file/profile` - Load only a credentials file with name `file`<sup>2</sup> and export the profile named `profile`.

`list` and `load profile` are a bit harder on GnuPG in that they decrypt all files, four at a time, concurrently. `load file/profile` will only load the specified file<sup>2</sup> and try to extract that profile.

---

##### Notes

<sup>1</sup>: Priority order is first `~/.aws/credentials` and then everything in `~/.aws/credentials.d/` in path-sorted order so that `~/.aws/credentials.d/99-something.ini` loads after `~/.aws/credentials.d/98-something.asc`.

<sup>2</sup>: To get filenames, the extension is removed and the path prefix as well, so `~/.aws/credentials.d/something.ini` and `~/.aws/credentials.d/something.asc` will both be found under `something/${profile}` when that is used.

**Comment by naftulikay** [comment]
Presently, it does not check for empty keys, as I thought that was handled with the type system, but we do need to check that `aws_access_key_id` and `aws_secret_access_key` are non-empty.

 - [x] Check that keys are not empty during INI parsing.

**Comment by naftulikay** [comment]
Profile listing is now working as expected but there appears to be a bug in the output ordering. The order within the collection must be:

- `~/.aws/credentials` **ALWAYS FIRST**
- `~/.aws/credentials.d/00-something.ini`
- `~/.aws/credentials.d/99-other.ini`

However, when creating a _view_ of this, the view must be _reversed_, such that `~/.aws/credentials` should be the _last_ thing inserted, followed by `~/.aws/credentials.d/00-something.ini` and finally `~/.aws/credentials.d/99-other.ini`.

Basically, go through sorted order of contents in `~/.aws/credentials.d/` and insert in that sorted order, `00.ini`, `01.ini`, etc. such that each subsequent value supersedes the previous value, and then finally, at the very end, insert `~/.aws/credentials`, such that _it_ supersedes everything else.

Best thing to do here is to write test cases, as this is dumb and stupid and hard to understand.

**Comment by naftulikay** [comment]
Updated PR description, superseding all of the comments above.

**Comment by naftulikay** [comment]
### TODO

 - [ ] Refactor `AwsProfileLookup` to do the actual business logic of lazy-loading.
   - e.g. `pub fn get_lazy(&mut self, pattern: &str) -> Result<Option<AwsProfile>, Error>`

Internally, `get_lazy` should copy the business logic above, but not repeat loads for files already in the index. This likely means that we should either get rid of one of `AwsCredentials` and `AwsProfileLookup` and provide internal methods for loading the root credentials file, files by a file stem, all files not currently found in the index, and a method for loading all files for the `list` use-case.

**Comment by naftulikay** [comment]
### FIXME

Add MIT and Apache 2.0 licenses.
